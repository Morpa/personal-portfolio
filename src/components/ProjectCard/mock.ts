export default [
  {
    name: 'My project',
    description: 'My descriptio',
    platform: 'web',
    stack: 'React, styledâ€‘components',
    source: 'https://github.com/',
    demo: 'https://mysite.com/'
  }
]
