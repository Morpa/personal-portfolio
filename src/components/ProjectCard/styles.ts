import styled, { css } from 'styled-components'
import { darken } from 'polished'
import { motion } from 'framer-motion'
import media from 'styled-media-query'

import * as ButtonStyles from 'components/Button/styles'
import * as RibbonStyles from 'components/Ribbon/styles'

export const Wrapper = styled.article`
  ${({ theme }) => css`
    position: relative;
    display: flex;
    flex-direction: column;
    width: 100%;
    height: 100%;
    background-color: ${theme.colors.white};
    border-radius: ${theme.border.radius};

    ${media.lessThan('large')`
      ${RibbonStyles.Wrapper} {
        right: 0;
        &::before {
          display: none;
        }
      }
  `}
  `}
`

export const ImageBox = styled.div`
  height: 24rem;
  width: 100%;
  background: #f6f7f8;
  background-image: linear-gradient(
    to right,
    #f6f7f8 0%,
    #edeef1 20%,
    #f6f7f8 40%,
    #f6f7f8 100%
  );
  background-size: 80rem 14rem;
  animation: placeholderShimmer 1s linear infinite forwards;

  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  @keyframes placeholderShimmer {
    0% {
      background-position: -40rem 0;
    }

    100% {
      background-position: 40rem 0;
    }
  }
`

export const Content = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    position: relative;
    height: 100%;
    margin: ${theme.spacings.xsmall};
  `}
`

export const Info = styled.div`
  max-width: calc(100% - 2.5rem);
`

export const Title = styled.h3`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.xlarge};
    line-height: ${theme.font.sizes.medium};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.black};
  `}
`

export const ButtonsWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    justify-content: flex-end;
    margin-top: ${theme.spacings.small};
  `}
`
export const Live = styled(motion.div)`
  ${({ theme }) => css`
    ${ButtonStyles.Wrapper} {
      background: ${theme.colors.mainBg};
      font-size: ${theme.font.sizes.xlarge};
      margin-right: ${theme.spacings.xxsmall};

      svg {
        width: 3rem;
      }

      &:hover {
        background: ${darken(0.1, theme.colors.secondary)};
      }
    }
  `}
`
export const Git = styled(motion.div)`
  ${({ theme }) => css`
    ${ButtonStyles.Wrapper} {
      background: ${theme.colors.mainBg};
      font-size: ${theme.font.sizes.xlarge};

      svg {
        width: 3rem;
      }

      &:hover {
        background: ${darken(0.1, theme.colors.secondary)};
      }
    }
  `}
`
