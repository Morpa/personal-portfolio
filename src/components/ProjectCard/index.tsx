import { useEffect, useState } from 'react'
import { Github } from '@styled-icons/boxicons-logos'
import { Computer, Add, Close } from '@styled-icons/material-outlined'
import { MotionProps } from 'framer-motion'

import Ribbon, { RibbonColors, RibbonSizes } from 'components/Ribbon'
import Button from 'components/Button'
import Heading from 'components/Heading'

import * as S from './styles'

const animationSpring: MotionProps = {
  whileHover: { y: -4 },
  whileTap: {
    y: 0
  },
  transition: {
    type: 'spring',
    stiffness: 250,
    damping: 14
  }
}

export type ProjectCardProps = {
  title: string
  img: string
  url?: string
  techs?: string[]
  description?: string
  github: string
  ribbon?: React.ReactNode
  ribbonColor?: RibbonColors
  ribbonSize?: RibbonSizes
}

const ProjectCard = ({
  title,
  img,
  url,
  techs,
  description,
  github,
  ribbon,
  ribbonColor = 'primary',
  ribbonSize = 'normal'
}: ProjectCardProps) => {
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    const handleKeyUp = ({ key }: KeyboardEvent) => {
      key === 'Escape' && setIsOpen(false)
    }

    window.addEventListener('keyup', handleKeyUp)

    return () => window.removeEventListener('keyup', handleKeyUp)
  }, [])

  return (
    <S.Wrapper>
      {!!ribbon && (
        <Ribbon color={ribbonColor} size={ribbonSize}>
          {ribbon}
        </Ribbon>
      )}
      <S.ImageBox>
        <img src={img} alt={title} />
      </S.ImageBox>
      <S.Content>
        <S.Info>
          <Heading lineLeft color="black" lineColor="secondary">
            {title}
          </Heading>
        </S.Info>

        <S.ButtonsWrapper>
          {!!url && (
            <S.Live {...animationSpring}>
              <Button
                icon={<Computer />}
                size="small"
                as="a"
                href={url}
                aria-label="live"
              >
                Live
              </Button>
            </S.Live>
          )}
          <S.More {...animationSpring}>
            <Button
              icon={<Add />}
              size="small"
              aria-label="open modal"
              onClick={() => setIsOpen(true)}
            />
          </S.More>
          <S.Git {...animationSpring}>
            <Button
              icon={<Github />}
              size="small"
              as="a"
              href={github}
              aria-label="github"
            >
              Github
            </Button>
          </S.Git>
        </S.ButtonsWrapper>
      </S.Content>

      <S.Modal isOpen={isOpen} aria-label="modal" aria-hidden={!isOpen}>
        <S.Close
          role="button"
          aria-label="close modal"
          onClick={() => setIsOpen(false)}
        >
          <Close size={40} />
        </S.Close>

        <S.ModalContent>
          {!!techs && (
            <>
              <Heading lineBottom lineColor="primary" size="huge">
                Techs
              </Heading>

              <S.TechsWrapper>
                {techs.map((tech, index) => (
                  <S.Tech key={`tech-${index}`}> {tech}</S.Tech>
                ))}
              </S.TechsWrapper>
            </>
          )}

          {!!description && (
            <>
              <Heading lineBottom lineColor="primary" size="huge">
                Description
              </Heading>
              <S.Description>{description}</S.Description>
            </>
          )}
        </S.ModalContent>
      </S.Modal>
    </S.Wrapper>
  )
}

export default ProjectCard
