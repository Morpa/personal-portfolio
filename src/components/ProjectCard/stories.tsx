import { Story, Meta } from '@storybook/react/types-6-0'
import ProjectCard, { ProjectCardProps } from '.'

export default {
  title: 'ProjectCard',
  component: ProjectCard,
  args: {
    title: 'Population Zero',
    img: 'https://source.unsplash.com/user/willianjusten/300x140',
    github: 'https://git.com',
    url: 'https://site.com'
  },
  parameters: {
    backgrounds: {
      default: 'port-dark'
    }
  }
} as Meta

export const Default: Story<ProjectCardProps> = (args) => (
  <ProjectCard {...args} />
)

export const WithRibbon: Story<ProjectCardProps> = (args) => (
  <div style={{ width: '30rem' }}>
    <ProjectCard {...args} />
  </div>
)

WithRibbon.args = {
  ribbon: 'Mais recente',
  ribbonSize: 'small',
  ribbonColor: 'primary'
}
