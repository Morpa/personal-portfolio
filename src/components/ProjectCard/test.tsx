import { fireEvent, screen } from '@testing-library/react'

import { renderWithTheme } from 'utils/tests/helpers'

import ProjectCard from '.'

const props = {
  title: 'Population Zero',
  img: '/img/games/red-dead-img.jpg',
  github: 'https://git.com',
  url: 'https://site.com',
  ribbon: 'NEW',
  description:
    'Lorem ipsum dolor sit amet consectetur adipisicing elit. Voluptasipsam expedita unde, cupiditate eos temporibus veritatis minima etmaiores maxime mollitia saepe ab, incidunt quod, quo doloribusdelectus omnis ratione.',
  techs: ['React', 'Next']
}

describe('<ProjectCard />', () => {
  it('should render ProjectCard correctly', () => {
    const { container } = renderWithTheme(<ProjectCard {...props} />)

    expect(screen.getByRole('heading', { name: props.title })).toHaveStyle({
      color: '#030517'
    })

    expect(screen.getByRole('img', { name: props.title })).toHaveAttribute(
      'src',
      props.img
    )

    expect(screen.getByText(/github/i)).toBeInTheDocument()

    expect(screen.getByText(/live/i)).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render without live button', () => {
    renderWithTheme(<ProjectCard {...props} url="" />)

    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()

    expect(screen.getByRole('img', { name: props.title })).toHaveAttribute(
      'src',
      props.img
    )

    expect(screen.getByText(/github/i)).toBeInTheDocument()

    expect(screen.queryByText(/live/i)).not.toBeInTheDocument()
  })

  it('should render Ribbon', () => {
    renderWithTheme(
      <ProjectCard
        {...props}
        ribbon="My Ribbon"
        ribbonColor="secondary"
        ribbonSize="small"
      />
    )
    const ribbon = screen.getByText(/my ribbon/i)

    expect(ribbon).toHaveStyle({ backgroundColor: '#3CD3C1' })

    expect(ribbon).toHaveStyle({ height: '2.6rem', fontSize: '1.2rem' })

    expect(ribbon).toBeInTheDocument()
  })

  it('should handle open modal', () => {
    renderWithTheme(<ProjectCard {...props} />)

    const modal = screen.getByLabelText('modal')

    expect(modal.getAttribute('aria-hidden')).toBe('true')

    expect(modal).toHaveStyle({ opacity: 0, pointerEvents: 'none' })

    fireEvent.click(screen.getByRole('button', { name: /open modal/i }))

    expect(modal.getAttribute('aria-hidden')).toBe('false')

    expect(modal).toHaveStyle({ opacity: 1 })
  })

  it('should handle close modal when overlay or button is clicked', () => {
    renderWithTheme(<ProjectCard {...props} />)

    const modal = screen.getByLabelText('modal')

    fireEvent.click(screen.getByRole('button', { name: /open modal/i }))

    fireEvent.click(screen.getByRole('button', { name: /close modal/i }))

    expect(modal.getAttribute('aria-hidden')).toBe('true')

    expect(modal).toHaveStyle({ opacity: 0 })
  })

  it('should handle close modal when ESC button is pressed', () => {
    const { container } = renderWithTheme(<ProjectCard {...props} />)

    const modal = screen.getByLabelText('modal')

    fireEvent.click(screen.getByRole('button', { name: /open modal/i }))

    fireEvent.keyUp(container, { key: 'Escape' })

    expect(modal.getAttribute('aria-hidden')).toBe('true')

    expect(modal).toHaveStyle({ opacity: 0 })
  })
})
