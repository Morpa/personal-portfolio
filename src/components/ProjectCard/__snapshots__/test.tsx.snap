// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectCard /> should render ProjectCard correctly 1`] = `
.c11 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c2 {
  position: absolute;
  top: 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 600;
  color: #FAFAFA;
  font-size: 1.4rem;
  padding: 0 2.4rem;
  height: 3.6rem;
  right: -2rem;
  background-color: #F231A5;
}

.c2::before {
  content: '';
  position: absolute;
  right: 0;
  border-style: solid;
  border-left-width: 0rem;
  border-right-color: transparent;
  border-bottom-color: transparent;
  border-bottom-width: 1rem;
}

.c2::before {
  top: 3.6rem;
  border-top-width: 1rem;
  border-right-width: 2rem;
}

.c2::before {
  border-left-color: #b20b6f;
  border-top-color: #b20b6f;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 3rem;
  font-size: 1.2rem;
}

.c10:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c10 svg {
  width: 1.5rem;
}

.c10 svg + span {
  margin-left: 0.8rem;
}

.c6 {
  color: #030517;
  font-weight: 400;
  padding-left: 0.8rem;
  border-left: 0.7rem solid #3CD3C1;
  font-size: 2.2rem;
}

.c17 {
  color: #FAFAFA;
  font-weight: 400;
  position: relative;
  margin-bottom: 3.2rem;
  font-size: 5.2rem;
}

.c17::after {
  position: absolute;
  left: 0;
  bottom: -1rem;
  content: '';
  width: 5rem;
  border-bottom: 0.5rem solid #F231A5;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  background-color: #FAFAFA;
  border-radius: 0.4rem;
}

.c3 {
  height: 24rem;
  width: 100%;
  background: #f6f7f8;
  background-image: linear-gradient( to right,#f6f7f8 0%,#edeef1 20%,#f6f7f8 40%,#f6f7f8 100% );
  background-size: 80rem 14rem;
  -webkit-animation: placeholderShimmer 1s linear infinite forwards;
  animation: placeholderShimmer 1s linear infinite forwards;
}

.c3 img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  position: relative;
  height: 100%;
  margin: 1.6rem;
}

.c5 {
  max-width: calc(100% - 2.5rem);
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-top: 2.4rem;
}

.c12 .c9 {
  background: #06092B;
  margin-right: 0.8rem;
}

.c12 .c9 svg {
  width: 3rem;
}

.c12 .c9:hover {
  background: #29b3a3;
}

.c8 .c9 {
  background: #06092B;
  font-size: 1.8rem;
  margin-right: 0.8rem;
}

.c8 .c9 > svg {
  width: 2.5rem;
}

.c8 .c9:hover {
  background: #29b3a3;
}

.c13 .c9 {
  background: #06092B;
  font-size: 1.8rem;
  margin-right: -0.6rem;
}

.c13 .c9 > svg {
  width: 2.5rem;
}

.c13 .c9:hover {
  background: #29b3a3;
}

.c14 {
  position: fixed;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  background: rgba(0,0,0,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 40;
  -webkit-transition: opacity 0.3s ease-in-out;
  transition: opacity 0.3s ease-in-out;
  padding: 0.8rem;
  opacity: 0;
  pointer-events: none;
}

.c15 {
  color: #FAFAFA;
  position: absolute;
  left: 0;
  top: 0;
  cursor: pointer;
  width: 100%;
  height: 100%;
  text-align: right;
}

.c16 {
  max-width: min(120rem,100%);
  max-height: 80rem;
}

.c20 {
  color: #FAFAFA;
}

.c18 {
  list-style: none;
  margin: 0.8rem 0;
  color: #FAFAFA;
}

.c19 {
  color: #FAFAFA;
}

.c19:before {
  content: 'âœ”';
  font-size: 2.0rem;
  color: #F231A5;
  margin-right: 0.8rem;
}

@media (min-width:1170px) {
  .c6 {
    font-size: 2.8rem;
  }
}

@media (max-width:1170px) {
  .c0 .c1 {
    right: 0;
  }

  .c0 .c1::before {
    display: none;
  }
}

@media (max-width:768px) {
  .c7 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media (max-width:768px) {
  .c12 .c9 svg {
    width: 2rem;
  }
}

@media (max-width:768px) {
  .c8 .c9 {
    font-size: 1.4rem;
  }

  .c8 .c9 svg {
    width: 2rem;
  }
}

@media (max-width:768px) {
  .c13 .c9 {
    font-size: 1.4rem;
    margin-right: -0.3rem;
  }

  .c13 .c9 svg {
    width: 2rem;
  }
}

<article
  class="c0"
>
  <div
    class="c1 c2"
    color="primary"
  >
    NEW
  </div>
  <div
    class="c3"
  >
    <img
      alt="Population Zero"
      src="/img/games/red-dead-img.jpg"
    />
  </div>
  <div
    class="c4"
  >
    <div
      class="c5"
    >
      <h2
        class="c6"
        color="black"
      >
        Population Zero
      </h2>
    </div>
    <div
      class="c7"
    >
      <div
        class="c8"
      >
        <a
          aria-label="live"
          class="c9 c10"
          href="https://site.com"
        >
          <svg
            aria-hidden="true"
            class="c11"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M20 18c1.1 0 1.99-.9 1.99-2L22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2H0v2h24v-2h-4zM4 6h16v10H4V6z"
            />
          </svg>
          <span>
            Live
          </span>
        </a>
      </div>
      <div
        class="c12"
      >
        <button
          aria-label="open modal"
          class="c9 c10"
        >
          <svg
            aria-hidden="true"
            class="c11"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
      </div>
      <div
        class="c13"
      >
        <a
          aria-label="github"
          class="c9 c10"
          href="https://git.com"
        >
          <svg
            aria-hidden="true"
            class="c11"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.026 2c-5.509 0-9.974 4.465-9.974 9.974 0 4.406 2.857 8.145 6.821 9.465.499.09.679-.217.679-.481 0-.237-.008-.865-.011-1.696-2.775.602-3.361-1.338-3.361-1.338-.452-1.152-1.107-1.459-1.107-1.459-.905-.619.069-.605.069-.605 1.002.07 1.527 1.028 1.527 1.028.89 1.524 2.336 1.084 2.902.829.091-.645.351-1.085.635-1.334-2.214-.251-4.542-1.107-4.542-4.93 0-1.087.389-1.979 1.024-2.675-.101-.253-.446-1.268.099-2.64 0 0 .837-.269 2.742 1.021a9.582 9.582 0 012.496-.336 9.554 9.554 0 012.496.336c1.906-1.291 2.742-1.021 2.742-1.021.545 1.372.203 2.387.099 2.64.64.696 1.024 1.587 1.024 2.675 0 3.833-2.33 4.675-4.552 4.922.355.308.675.916.675 1.846 0 1.334-.012 2.41-.012 2.737 0 .267.178.577.687.479C19.146 20.115 22 16.379 22 11.974 22 6.465 17.535 2 12.026 2z"
              fill-rule="evenodd"
            />
          </svg>
          <span>
            Github
          </span>
        </a>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-label="modal"
    class="c14"
  >
    <div
      aria-label="close modal"
      class="c15"
      role="button"
    >
      <svg
        aria-hidden="true"
        class="c11"
        fill="currentColor"
        focusable="false"
        height="40"
        viewBox="0 0 24 24"
        width="40"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
        />
      </svg>
    </div>
    <div
      class="c16"
    >
      <h2
        class="c17"
        color="white"
      >
        Techs
      </h2>
      <ul
        class="c18"
      >
        <li
          class="c19"
        >
           
          React
        </li>
        <li
          class="c19"
        >
           
          Next
        </li>
      </ul>
      <h2
        class="c17"
        color="white"
      >
        Description
      </h2>
      <p
        class="c20"
      >
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Voluptasipsam expedita unde, cupiditate eos temporibus veritatis minima etmaiores maxime mollitia saepe ab, incidunt quod, quo doloribusdelectus omnis ratione.
      </p>
    </div>
  </div>
</article>
`
